xbmc/cores/AudioEngine/Sinks/AESinkAUDIOTRACK.cpp

```diff
@@ -70,6 +72,134 @@ bool CAESinkAUDIOTRACK::HasAmlHD()
+      sType = "LINE_DIGITAL";
+    else if (iType == CJNIAudioDeviceInfo::TYPE_TELEPHONY)
+      sType = "TELEPHONY";
+    else if (iType == CJNIAudioDeviceInfo::TYPE_TV_TUNER)
+      sType = "TV_TUNER";
+    else if (iType == CJNIAudioDeviceInfo::TYPE_UNKNOWN)
+      sType = "UNKNOWN";
+    else if (iType == CJNIAudioDeviceInfo::TYPE_USB_ACCESSORY)
+      sType = "USB_ACCESSORY";
+    else if (iType == CJNIAudioDeviceInfo::TYPE_USB_DEVICE)
+      sType = "USB_DEVICE";
+    else if (iType == CJNIAudioDeviceInfo::TYPE_WIRED_HEADPHONES)
+      sType = "WIRED_HEADPHONES";
+    else if (iType == CJNIAudioDeviceInfo::TYPE_WIRED_HEADSET)
+      sType = "WIRED_HEADSET";
+    else
+      sType = iType;
+
+    CLog::Log(LOGDEBUG, "    id: %d, type: %s, isSink: %s, isSource: %s", dev.getId(), sType.c_str(), dev.isSink() ? "true" : "false", dev.isSource() ? "true" : "false");
+
+    std::ostringstream oss;
+    if (!dev.getChannelCounts().empty())
+    {
+      for (auto i : dev.getChannelCounts())
+        oss << i << " / ";
+      CLog::Log(LOGDEBUG, "    channel counts: %s", oss.str().c_str());
+    }
+    else
+      CLog::Log(LOGDEBUG, "    channel counts: any");
+
+    if (!dev.getChannelIndexMasks().empty())
+    {
+      oss.clear(); oss.str("");
+      for (auto i : dev.getChannelIndexMasks())
+        oss << i << " / ";
+      CLog::Log(LOGDEBUG, "    channel index masks: %s", oss.str().c_str());
+    }
+    else
+      CLog::Log(LOGDEBUG, "    channel index masks: any");
+
+    if (!dev.getChannelMasks().empty())
+    {
+      oss.clear(); oss.str("");
+      for (auto i : dev.getChannelMasks())
+        oss << i << " / ";
+      CLog::Log(LOGDEBUG, "    channel masks: %s", oss.str().c_str());
+    }
+    else
+      CLog::Log(LOGDEBUG, "    channel masks: any");
+
+    if (!dev.getEncodings().empty())
+    {
+      oss.clear(); oss.str("");
+      for (auto i : dev.getEncodings())
+      {
+        if (i == CJNIAudioFormat::ENCODING_PCM_16BIT)
+          oss << "PCM_16BIT" << " / ";
+        else if (i == CJNIAudioFormat::ENCODING_PCM_FLOAT)
+          oss << "PCM_FLOAT" << " / ";
+        else if (i == CJNIAudioFormat::ENCODING_AC3)
+          oss << "AC3" << " / ";
+        else if (i == CJNIAudioFormat::ENCODING_E_AC3)
+          oss << "E_AC3" << " / ";
+        else if (i == CJNIAudioFormat::ENCODING_DTS)
+          oss << "DTS" << " / ";
+        else if (i == CJNIAudioFormat::ENCODING_DTS_HD)
+          oss << "DTS_HD" << " / ";
+        else if (i == CJNIAudioFormat::ENCODING_DOLBY_TRUEHD)
+          oss << "DOLBY_TRUEHD" << " / ";
+        else if (i == CJNIAudioFormat::ENCODING_IEC61937)
+          oss << "IEC61937" << " / ";
+        else
+          oss << i << " / ";
+      }
+      CLog::Log(LOGDEBUG, "    encodings: %s", oss.str().c_str());
+    }
+    else
+      CLog::Log(LOGDEBUG, "    encodings: any");
+
+    if (!dev.getSampleRates().empty())
+    {
+      oss.clear(); oss.str("");
+      for (auto i : dev.getSampleRates())
```

Comment:
This is lot's of duplication. If you need a "list to string" better implement one.